clang/include/clang/Basic/LangOptions.h

```diff
@@ -968,10 +968,7 @@ class FPOptionsOverride {
       setAllowFPContractAcrossStatement();
   }
 
-  void setDisallowOptimizations() {
-    setFPPreciseEnabled(true);
-    setDisallowFPContract();
-  }
+  void setDisallowOptimizations() { setFPPreciseEnabled(true); }
```

Comment:
Actually, instead of just removing the call to `setDisallowFPContract()`, we should instead explicitly set the FP Contraction state to "on" (which is the default at `-O0`).  That is, we should call `setAllowFPContractWithinStatement()`.

In particular, if the user is compiling with `-ffast-math`, then the FP Contraction state will be `-ffp-contract=fast` (allowing FMA across multiple statements).  And if they applied `optnone`, then by removing the call to `setDisallowFPContract()`, we would mistakenly leave that inter-statement FMA enabled.  By calling `setAllowFPContractWithinStatement()` we are explicitly setting it the equivalent of the command-line option `-ffp-contract=on`, which is the desired state of `-O0` (and `optnone`).

Reply:
Actually this is done by the call to `setFPPreciseEnabled`: https://github.com/spavloff/llvm-project/blob/e518bbfed9eefbe6d9d952dd0354ee974f7a9305/clang/include/clang/Basic/LangOptions.h#L955-L969

Reply:
Ahhh..  I didn't realize.  Thanks!